### ISSUE TITLE
`textDocument/didClose` and `textDocument/didOpen` not fired on file rename changing case

### ISSUE THREAD

**Issue by JohnnyMorganz** [microsoft/vscode-languageserver-node#1186]
The specification states that
> Document renames should be signaled to a server sending a document close notification with the document’s old name followed by an open notification using the document’s new name

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_didRename

When a file is renamed to the same content but a change of case (e.g. `foo.txt` -> `Foo.txt`), these events *do not get fired*.
This causes a loss of synchronisation between the client and the server

Steps:
1) Create a dummy language server (e.g. from https://github.com/microsoft/vscode-extension-samples/tree/main/lsp-sample)
2) Enable trace logging (either messages or verbose)
3) Create a new file `foo.txt` and open it: observe a `textDocument/didOpen` request
4) Rename the file to `Foo.txt`.
Expected: a didClose and then a didOpen request.
Actual: nothing

Client + Server version: `8.0.2`

**Comment by dbaeumer** [comment]
This is triggered by VS Code. On Windows a case change doesn't really point to a new file. I guess this is why VS Code doesn't fire these events. Need to clarify with VS Code.

**Comment by JohnnyMorganz** [comment]
The main problem arises as the client will send a different URI when the document file case changes, so the server no longer finds the locally synced document when doing a naive equality comparison.
Should the server be doing this lowercase lookup itself then? Maybe this should be explicitly called out in the spec (unless it has already and I've probably missed it)

**Comment by dbaeumer** [comment]
I would rather argue that this should be managed by the client. If a client will use the new URI then I think it should send a close and open. If the client normalizes the URI (as VS Code does) then I think it is fine to not send a close and open.

**Comment by JohnnyMorganz** [comment]
Yeah, that's fair. I didn't notice that VSCode fully normalised the URI at first (looks like I wasn't using the normalised URIs to store the locally synced text document).
I think its best that I also do not rely on these events always firing on rename to change internal state.

Is the normalisation that VSCode does standardised? Parts of my server get their own paths from the filesystem, so if I have a file path to `Foo.txt`, but vscode uses the `foo.txt` normalised uri, if I want to compare 2 uri's its not equal

**Comment by dbaeumer** [comment]
VS Code uses the path in its form it was first opened to my knowledge (it doesn't call real path).

You are correct: if you want to match a file from disk with a URI you got from the client then you need to normalize it.
